zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java

```diff
@@ -402,7 +287,370 @@ public void broadcastReloadedNoteList() {
+      } catch (IOException e) {
+        LOG.error("Fail to reload notes from repository");
+      }
+    }
+
+    List<Note> notes = notebook.getAllNotes();
+    List<Map<String, Object>> notesInfo = new LinkedList<>();
+    for (Note note : notes) {
+      boolean isNotebookRunning = false;
+      Map<String, Object> info = new HashMap<>();
+
+      if (hideHomeScreenNotebookFromList && note.id().equals(homescreenNotebookId)) {
+        continue;
+      }
+
+      String CRON_TYPE_NOTEBOOK_KEYWORD = "cron";
+      info.put("notebookId", note.id());
+      String notebookName = note.getName();
+      if (notebookName != null) {
+        info.put("notebookName", note.getName());
+      } else {
+        info.put("notebookName", "Note " + note.id());
+      }
+
+      if (note.getConfig().containsKey(CRON_TYPE_NOTEBOOK_KEYWORD) == true
+              && !note.getConfig().get(CRON_TYPE_NOTEBOOK_KEYWORD).equals("")) {
+        info.put("notebookType", "cron");
+      }
+      else {
+        info.put("notebookType", "normal");
+      }
+
+      long lastRunningUnixTime = 0;
+
+      List<Map<String, Object>> paragraphsInfo = new LinkedList<>();
+      for (Paragraph paragraph : note.getParagraphs()) {
+        if (paragraph.getStatus().isRunning() == true) {
+          isNotebookRunning = true;
+        }
+
+        Map<String, Object> paragraphItem = getParagraphPacketItem(paragraph);
+
+        lastRunningUnixTime = getUnixTimeLastRunParagraph(paragraph);
+
+        paragraphsInfo.add(paragraphItem);
+      }
+
+      // Interpreter is set does not exist.
+      String interpreterGroupName = null;
+      if (note.getNoteReplLoader().getInterpreterSettings() != null
+              && note.getNoteReplLoader().getInterpreterSettings().size() >= 1) {
+        interpreterGroupName = note.getNoteReplLoader().getInterpreterSettings().get(0).getGroup();
+      }
+
+      // notebook json object root information.
+      info.put("interpreter", interpreterGroupName);
+      info.put("isRunningJob", isNotebookRunning);
+      info.put("unixTimeLastRun", lastRunningUnixTime);
+      info.put("paragraphs", paragraphsInfo);
+      notesInfo.add(info);
+    }
+    return notesInfo;
+  }
+
+
+  public List<Map<String, Object>> generateUpdateNotebooksJobInfo(
+          boolean needsReload, long lastUpdateServerUnixTime) {
+    Notebook notebook = notebook();
+
+    ZeppelinConfiguration conf = notebook.getConf();
+    String homescreenNotebookId = conf.getString(ConfVars.ZEPPELIN_NOTEBOOK_HOMESCREEN);
+    boolean hideHomeScreenNotebookFromList = conf
+            .getBoolean(ConfVars.ZEPPELIN_NOTEBOOK_HOMESCREEN_HIDE);
+
+    if (needsReload) {
+      try {
+        notebook.reloadAllNotes();
+      } catch (IOException e) {
+        LOG.error("Fail to reload notes from repository");
+      }
+    }
+
+    List<Note> notes = notebook.getAllNotes();
+    List<Map<String, Object>> notesInfo = new LinkedList<>();
+    for (Note note : notes) {
```

Comment:
This method becomes quite long and a bit hard to understand. 
Would it be possible to refactor it a bit, extract few function with descriptive names and make it shorter and more clear, how do you think?

I was thinking about something along the lines:

```java
for (Note note: notes) {
  doSomethingPerNote();
  for (Paragraph p: paragraphs) {
    if (someReadableConditionName()) {
      continue; // skip this paragraph because of ... 'condition`
    }
    doSomethingUsefullPerParagraph();
    thenDoSomethingElse()
    if (!notebookWasUpdated) {
      continue;
    }
  }
}
```